# DUO.py

This notebook is the bases for the script_for_duos_playlist.py. It covers all steps which lead to the creation of a new duos playlist for my girlfriend and me.

The notebook was used for experimentation and building the steps to arrive at a DUO.py playlist that was enjoyable to both users.

I wanted to create a playlist that was different from Spotify's 'Duo Mix', which didn't really seem to grasp the music tastes of my girlfriend and me and was somehow disappointing.

The playlist we had hoped for wasn't available so I decided to create one, that could be updated periodically and would incorporate some features:
- include some all time favorites as well as songs we recently both enjoyed
- include new songs based on our favorite songs
- don't include too many songs from the same artist or genre
- provide a listening experience that is based on mutually liked music not just a combination of our individual tastes

Right now the workflow accomplishes this although it is not yet perfect. I intend to add machine learning to further "filter" songs based on what we would like to have in our personal playlist. This howoever requires some data that is currently collected as we enjoy and discard songs from the weekly playlist.

## Imports

The most important imports are spotipy, pandas, numpy and my own small library 'spotifuncs'. Here I use the wildcard import for experimentation, the scripts only imports necessary functions. spotifuncs itself uses pandas, sklearn, spotipy.

In [1]:
import spotipy
from spotipy.oauth2 import SpotifyClientCredentials
import spotipy.util as util
from pathlib import Path
import pandas as pd
import numpy as np
from spotifuncs import *

In [2]:
path = Path("C:/Users/ms101/OneDrive/DataScience_ML/projects/spotify_app")

## Setting the Credentials 

I stored the credentials and usernames in a .txt that was not uploaded to GitHub to ensure safety and avoid showing sensible information in my notebooks. 

Here I simply read the lines within the .txt files to retrieve the necessary information

In [3]:
with open(path / "client_s.txt") as f:
    content = f.readlines()
content = [x.strip() for x in content]

client_id = content[0]
client_secret = content[1]

In [4]:
with open(path / "usernames.txt") as f:
    usernames = f.readlines()
usernames = [x.strip() for x in usernames]

username1 = usernames[0]
username2 = usernames[1]

## App Scope

For the following code to work the scope needs to be quite extensive as I am retrieving a lot of user data and modifying a collaborative playlist (which the SpotifyAPI automatically views as private). As the only users are my girlfriend and me this was okay. 

**If others should choose to use my app a thorough explanation concerning the use of their data and the abilities of the app are paramount.**


To view all available scopes click here: https://developer.spotify.com/documentation/general/guides/scopes/

In [5]:
scope = "user-library-read user-read-recently-played user-top-read playlist-modify-public playlist-read-private playlist-read-collaborative"

redirect_uri = "https://developer.spotify.com/dashboard/applications/4a4e029d299a4241873db8300038bf0a"

client_credentials_manager = SpotifyClientCredentials(client_id=client_id, 
                                                      client_secret=client_secret)


## Authenticate

I reduced the authentication process to a function that can be found within the spotifuncs library. I describe it in more detail in my [Medium post](https://towardsdatascience.com/using-python-to-refine-your-spotify-recommendations-6dc08bcf408e).

In [6]:
sp_m = authenticate(redirect_uri, client_credentials_manager, username1, scope, client_id, client_secret)
sp_t = authenticate(redirect_uri, client_credentials_manager, username2, scope, client_id, client_secret)

## Retrieving the Data

I am retrieving quite a few dictonaries that end up in multiple dataframes. I packaged the process into a function because I have to repeat it. I decided not to integrate it into spotifuncs, as the function is very specific and subject to changes.

The first important piece of data are all the top tracks of the user, which I combine into a complete list of top tracks that cover everything from short to long term favorites. Notice that I am sampling only 15 songs from the long term top tracks and doing so **without** setting a random seed to **not** get the same results everytime the code is run. Always using the full list of long term favorites in the process would lead to too much repition over the course of multiple weeks and thus playlists.

The second important piece are the top artists, which like the tracks are retrieved for all time frames. The artists are important for the filtering process later on.

Lastly I am also retrieving the last 50 tracks a user saved. 50 is the upper limit here, which is unfortunate as this really limits the use of the data.

In [7]:
def get_dfs(sp):
    ##queries
    #user top tracks
    top_tracks_short = sp.current_user_top_tracks(limit = 50,offset=0,time_range='short_term')
    top_tracks_med = sp.current_user_top_tracks(limit = 50,offset=0,time_range='medium_term')
    top_tracks_long = sp.current_user_top_tracks(limit = 50,offset=0,time_range='long_term')
    
    #combine the top_tracks
    top_tracks_short_df = append_audio_features(create_df_top_songs(top_tracks_short),sp)
    top_tracks_med_df = append_audio_features(create_df_top_songs(top_tracks_med),sp)
    top_tracks_long_df = append_audio_features(create_df_top_songs(top_tracks_long),sp)
    #sample from long-term top tracks to introduce more randomness and avoid having the same artists
    top_tracks_long_df = top_tracks_long_df.sample(n = 15)
    top_tracks_df = pd.concat([top_tracks_short_df,top_tracks_med_df,top_tracks_long_df]).drop_duplicates().reset_index(drop = True)
        
    #user top artists
    top_artists_long = sp.current_user_top_artists(limit = 50, time_range = "long_term")
    top_artists_med = sp.current_user_top_artists(limit = 50, time_range = "medium_term")
    top_artists_short = sp.current_user_top_artists(limit = 50, time_range = "short_term")
    
    artists_short_df = top_artists_from_API(top_artists_short)
    artists_med_df = top_artists_from_API(top_artists_med)
    artists_long_df = top_artists_from_API(top_artists_long)
    artists_df = pd.concat([artists_short_df,artists_med_df,artists_long_df])
    artists_df["genres"] = artists_df["genres"].apply(lambda x: ",".join(x))
    artists_df.drop_duplicates().reset_index(drop = True)
    
    #user saved tracks
    user_saved_tracks = sp.current_user_saved_tracks(limit = 50)
    saved_tracks_df = create_df_saved_songs(user_saved_tracks)
    
        
    return top_tracks_df,artists_df,saved_tracks_df

In [8]:
top_tracks_m, artists_m, saved_tracks_m = get_dfs(sp_m)

In [9]:
top_tracks_t, artists_t, saved_tracks_t = get_dfs(sp_t)

In [10]:
artists_t

name                      id  \
0        Emancipator  6HCnsY0Rxi3cg53xreoAIm   
1            Chelina  3XQZW9cuoDf7JhPbr99bXD   
2   Element Of Crime  3FweAJRBCbUOGR6jULfaRi   
3         Mac Miller  4LLpKhyESsyAXpc4laK94U   
4    Kid Francescoli  2G7QgTep5IsJHGHm1hXygD   
..               ...                     ...   
45            Rakede  4soVkCNrRQccCv4Nohz273   
46            257ers  6ihLfpY3cmdGyWEnItn30w   
47             MEUTE  1z5xbcOeFRQXBVDpvRPh8H   
48         Max Herre  7IpWQKu80qQvyer3LO6SW3   
49            Hozier  2FXC3k01G6Gw61bmprjgqS   

                                               genres  popularity  \
0          downtempo,electronica,livetronica,trip hop          62   
1                                         amharic pop          13   
2    german indie,german pop,german rock,liedermacher          51   
3                          hip hop,pittsburgh rap,rap          85   
4   french indie pop,french indietronica,new frenc...          60   
..                                                ...         ...   
45                                    hamburg hip hop          40   
46  antideutsche,deep german hip hop,german hip ho...          62   
47        german dance,hamburg electronic,livetronica          53   
48   german alternative rap,german hip hop,german pop          53   
49            irish singer-songwriter,modern rock,pop          80   

                                      uri  
0   spotify:artist:6HCnsY0Rxi3cg53xreoAIm  
1   spotify:artist:3XQZW9cuoDf7JhPbr99bXD  
2   spotify:artist:3FweAJRBCbUOGR6jULfaRi  
3   spotify:artist:4LLpKhyESsyAXpc4laK94U  
4   spotify:artist:2G7QgTep5IsJHGHm1hXygD  
..                                    ...  
45  spotify:artist:4soVkCNrRQccCv4Nohz273  
46  spotify:artist:6ihLfpY3cmdGyWEnItn30w  
47  spotify:artist:1z5xbcOeFRQXBVDpvRPh8H  
48  spotify:artist:7IpWQKu80qQvyer3LO6SW3  
49  spotify:artist:2FXC3k01G6Gw61bmprjgqS  

[108 rows x 5 columns]

In [11]:
artists_m

name                      id  \
0                Rammstein  6wWVKhxIU2cEi0K81v7HvP   
1            Frank Sinatra  1Mxqyy3pSjf8kZZL4QVxS0   
2                      SCH  2kXKa3aAFngGz2P4GjG5w2   
3                   Geegun  5W7N6u4EjCEMKj7bDyzPEC   
4             Joyner Lucas  6C1ohJrd5VydigQtaGy5Wa   
..                     ...                     ...   
45              Chris Rock  36eSjIksD6fehqxyDUHDA3   
46                Wardruna  0NJ6wlOAsAJ1PN4VRdTPKA   
47                 Carnage  7CCjtD0hCK005Bvg2WG1a7   
48               Motörhead  1DFr97A9HnbV3SKTJFu62M   
49  Thirty Seconds To Mars  0RqtSIYZmd4fiBKVFqyIqD   

                                               genres  popularity  \
0   alternative metal,german metal,industrial,indu...          79   
1               adult standards,easy listening,lounge          89   
2   french hip hop,pop urbaine,rap francais,rap ma...          77   
3   russian dance,russian dance pop,russian hip ho...          57   
4                  boston hip hop,hip hop,pop rap,rap          77   
..                                                ...         ...   
45                                black comedy,comedy          48   
46    medieval folk,nordic folk,rune folk,viking folk          62   
47  edm,electro house,electronic trap,pop rap,rap,...          64   
48        album rock,hard rock,metal,rock,speed metal          69   
49              modern rock,pop punk,post-grunge,rock          70   

                                      uri  
0   spotify:artist:6wWVKhxIU2cEi0K81v7HvP  
1   spotify:artist:1Mxqyy3pSjf8kZZL4QVxS0  
2   spotify:artist:2kXKa3aAFngGz2P4GjG5w2  
3   spotify:artist:5W7N6u4EjCEMKj7bDyzPEC  
4   spotify:artist:6C1ohJrd5VydigQtaGy5Wa  
..                                    ...  
45  spotify:artist:36eSjIksD6fehqxyDUHDA3  
46  spotify:artist:0NJ6wlOAsAJ1PN4VRdTPKA  
47  spotify:artist:7CCjtD0hCK005Bvg2WG1a7  
48  spotify:artist:1DFr97A9HnbV3SKTJFu62M  
49  spotify:artist:0RqtSIYZmd4fiBKVFqyIqD  

[136 rows x 5 columns]

### Finding common artists

I am finding common artists (of the 2 users) to later filter there top songs by artists. 
The logic behind this is the following:

A track might be only among one users "top tracks", it may however be by an artist both users enjoy. In that case the track is a good candidate for the duos playlist as both users will probably enjoy it, yet it might be a new discovery for one of them. If both already know and like it: still a good fit for the playlist!

In [12]:
common_artists = dataframe_difference(artists_m,artists_t, which = "both")
common_artists

C:\Users\ms101\Anaconda3\envs\spotify_app\lib\site-packages\pandas\core\frame.py:4163: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


name                      id  \
0       Frank Sinatra  1Mxqyy3pSjf8kZZL4QVxS0   
1        Joyner Lucas  6C1ohJrd5VydigQtaGy5Wa   
2              G-Eazy  02kJSzxNuaWGqwubyUba0Z   
3        Modeselektor  2jYMYP2SVifgmzNRQJx3SJ   
4     Kid Francescoli  2G7QgTep5IsJHGHm1hXygD   
5               Seeed  5ISjkNS17JpCwiFtW80lpV   
6          The Weeknd  1Xyo4u8uXC1ZmMpatF05PJ   
7   Christian Löffler  3tSvlEzeDnVbQJBTkIA6nO   
8            Monolink  2I4hRNCYkPKJQlkoEZKjYx   
9          Egor Kreed  2KoLmBXwsgMkfAvoPBlPmb   
10               ELIF  65AzRSW0jKSs0WtttEXrOw   
11        KitschKrieg  5tHiL8SKSaZGMBUPIiSmX4   
12     Arctic Monkeys  7Ln80lUS6He07XvHI8qqHH   
13             Eminem  7dGJo4pcD2V6oG8kP0tJRR   
14           Kontra K  4lDiJcOJ2GLCK6p9q5BgfK   

                                               genres  popularity  \
0               adult standards,easy listening,lounge          89   
1                  boston hip hop,hip hop,pop rap,rap          77   
2   hip hop,indie pop rap,oakland hip hop,pop rap,rap          83   
3   alternative dance,electronica,microhouse,minim...          50   
4   french indie pop,french indietronica,new frenc...          60   
5             german hip hop,german pop,german reggae          65   
6          canadian contemporary r&b,canadian pop,pop          94   
7   electronica,hamburg electronic,microhouse,mini...          62   
8                                       organic house          62   
9                         russian hip hop,russian pop          66   
10      frauenrap,german pop,german singer-songwriter          68   
11                     german hip hop,hamburg hip hop          70   
12  garage rock,modern rock,permanent wave,rock,sh...          86   
13                        detroit hip hop,hip hop,rap          94   
14                    christlicher rap,german hip hop          74   

                                      uri  
0   spotify:artist:1Mxqyy3pSjf8kZZL4QVxS0  
1   spotify:artist:6C1ohJrd5VydigQtaGy5Wa  
2   spotify:artist:02kJSzxNuaWGqwubyUba0Z  
3   spotify:artist:2jYMYP2SVifgmzNRQJx3SJ  
4   spotify:artist:2G7QgTep5IsJHGHm1hXygD  
5   spotify:artist:5ISjkNS17JpCwiFtW80lpV  
6   spotify:artist:1Xyo4u8uXC1ZmMpatF05PJ  
7   spotify:artist:3tSvlEzeDnVbQJBTkIA6nO  
8   spotify:artist:2I4hRNCYkPKJQlkoEZKjYx  
9   spotify:artist:2KoLmBXwsgMkfAvoPBlPmb  
10  spotify:artist:65AzRSW0jKSs0WtttEXrOw  
11  spotify:artist:5tHiL8SKSaZGMBUPIiSmX4  
12  spotify:artist:7Ln80lUS6He07XvHI8qqHH  
13  spotify:artist:7dGJo4pcD2V6oG8kP0tJRR  
14  spotify:artist:4lDiJcOJ2GLCK6p9q5BgfK

### Last weeks playlist

In order to avoid encountering the same songs two weeks in a row, which is very likely as short and medium term top tracks won't have changed much, last weeks playlist is read from 'Playlist.csv'. This file was just an empty csv file the first time the code was run. But at the end of the playlist creation process the created playlist is saved in that csv file, so it contains last weeks playlist.

The code that creates the playlist (script_for_duos_playlist.py) does the same every time it is run.

In [13]:
last_week_duo = pd.read_csv(path/"Playlist.csv", index_col = 0)
last_week_duo

track_name  \
0                           Who - Single Version   
1                                     Coco L'Eau   
2                                 A Child's Tale   
3                                       Bläulich   
4                                    Lambo Lambo   
5                             Come Waltz With Me   
6                         The Beautiful & Damned   
7                      KANN DAS BITTE SO BLEIBEN   
8                         I Mean It (feat. Remo)   
9                                            Neo   
10                                        Wealth   
11                                    Had Enough   
12                                       Не могу   
13      Why'd You Only Call Me When You're High?   
14                                       The End   
15                                           505   
16                                    What A Day   
17                                  Five Minutes   
18                                When I'm Small   
19                                  C'est La Vie   
20                                   Ghostkeeper   
21                                     Колхозник   
22          Silver Tongue Devil (feat. Shenseea)   
23                                         sugar   
24            Take Some Time - Emancipator Remix   
25               Born Slippy (Nuxx) - Radio Edit   
26                  All in You (feat. Anna Kova)   
27                                      Big Iron   
28                                    Oh Brother   
29                      You Too Can Be a Dreamer   
30                                   Eine Minute   
31                            Go Hard or Go Home   
32                                   Hell & Back   
33                                     Suspirium   
34                                     Bird Song   
35                         Your Body Is a Weapon   
36                                     Sunnedeck   
37                                       BIG BOI   
38       Up Past The Nursery - Ivan Smagghe Edit   
39                            Only Ones Who Know   
40                                 Tanz für mich   
41                                       Houdini   
42                                Ich mach alles   
43  Clap Your Hands - Solomun Remix / Radio Edit   
44                          Ill Mind of Hopsin 8   
45                                 New President   

                                            album                track_id  \
0                                             Who  0CjBORMsmiQNe3vPDcNIvk   
1                                      Coco L'Eau  5weiiB92gNV7QHFYQXqxZ8   
2                                       Phantasma  2JNa5xzODo5tiHDIvLPpGt   
3                                     Treppenhaus  2WRTnY0slmFgWcrmEr8dPj   
4                                     KitschKrieg  7oqvRZNv4dUV8CgQWtIAMe   
5                       Reprise Rarities (Vol. 1)  1SfTz9jOsbY9iS3YLQt0CK   
6                          The Beautiful & Damned  2WWruw7ul9N7eqoHELyMc2   
7                                           NACHT  45HOck8XCgrSlVUQHHOHMz   
8                             These Things Happen  6jmTHeoWvBaSrwWttr8Xvu   
9                                            Mare  1xm2dkD3qpHdw9h7YTDozm   
10                                       Who Else  5aOlYhQsp75cgPov4yjWIe   
11                      Everything's Strange Here  26Td0OU1dTm3Je3ynSLC1a   
12                                 Что они знают?  72aW82jpOZ6L0Ran6cZK2q   
13                                             AM  086myS9r57YsLbJpU0TgK9   
14                                        The End  0lwUE9DGjdbbuwbPbxZyzw   
15                      Favourite Worst Nightmare  0BxE4FqsDD1Ot4YuBXwAPp   
16                                        BAM BAM  3Zexl3b8tdk5csqUHgTPnZ   
17                                   Five Minutes  3pJnvBbwIvuwebHHTYzoR3   
18                                  Eyelid Movies  3498wF96LsgVgMkGmJzJOC   
19                                       Part Two  5V

## Creating the Playlist

The creation of the playlist is the main goal of the task and of the project. It requires a couple of steps to 'assemble all building blocks' that make up the DUO.py playlist. THe building blocks are:

1. Common top tracks, that were not in last weeks playlists
2. A sample of each users top tracks that are most similar to the other users top tracks
3. A sample of each users top tracks from an artists both users like (one of their top artists)
4. A sample of the songs saved by users
5. A recommended track (through spotify and additional filtering) for every track that was added to the playlist in step 1.-4.

### Common top tracks

The playlist is initiate by common top tracks, that did not appear in last weeks playlist already. As these songs are both users' favorites they should enjoy them.

In [14]:
dataframe_difference(top_tracks_m,top_tracks_t,which = "both")

C:\Users\ms101\Anaconda3\envs\spotify_app\lib\site-packages\pandas\core\frame.py:4163: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


track_name        album                track_id        artist  \
0    Who - Single Version          Who  0CjBORMsmiQNe3vPDcNIvk  Modeselektor   
1              Coco L'Eau   Coco L'Eau  5weiiB92gNV7QHFYQXqxZ8    Egor Kreed   
2          A Child's Tale    Phantasma  2JNa5xzODo5tiHDIvLPpGt      Bukahara   
3  Huldra - Other Version       Huldra  569gNjph2g07MmjtMm6vKm         Gidge   
4                Bläulich  Treppenhaus  2WRTnY0slmFgWcrmEr8dPj    Apache 207   
5                  Wealth     Who Else  5aOlYhQsp75cgPov4yjWIe  Modeselektor   

   duration  popularity  danceability  energy  key  loudness  mode  \
0    207988          43         0.760   0.910    7    -8.472     1   
1    130050          64         0.837   0.663    3    -5.071     0   
2    179693          47         0.668   0.398    0   -10.539     0   
3    490000          38         0.772   0.514   10   -11.961     0   
4    196213          69         0.790   0.704   10    -7.935     0   
5    247218          41         0.765   0.452    1   -12.346     1   

   speechiness  acousticness  instrumentalness  liveness  valence    tempo  
0       0.0484        0.0442          0.006620    0.1080    0.392  142.982  
1       0.0497        0.0949          0.009540    0.1750    0.466  100.006  
2       0.0460        0.8650          0.010200    0.2170    0.961  166.874  
3       0.0754        0.1480          0.906000    0.0934    0.103  118.015  
4       0.4170        0.0690          0.000658    0.1130    0.212  154.007  
5       0.0856        0.0204          0.000057    0.0727    0.331  137.972

In [16]:
common_songs = dataframe_difference(top_tracks_m,top_tracks_t,which = "both")
new_playlist_df = common_songs[~common_songs["track_id"].isin(last_week_duo["track_id"])]
new_playlist_df

C:\Users\ms101\Anaconda3\envs\spotify_app\lib\site-packages\pandas\core\frame.py:4163: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


track_name   album                track_id artist  duration  \
3  Huldra - Other Version  Huldra  569gNjph2g07MmjtMm6vKm  Gidge    490000   

   popularity  danceability  energy  key  loudness  mode  speechiness  \
3          38         0.772   0.514   10   -11.961     0       0.0754   

   acousticness  instrumentalness  liveness  valence    tempo  
3         0.148             0.906    0.0934    0.103  118.015

### Users top tracks that are similar to each other.

The next building block consist of songs from both users top tracks that are most similar to on another on the level of audio features.

For this task the top tracks unique to each user are extracted and a similarity matrix is computed. The similarity is based on the audio features (excluding 'key' and 'mode') and computed via cosine similarity (see spotifuncs).
From this matrix the 30 highest similarity scores and corresponding indeces are extracted. The songs corresponding to these indeces are put into a dataframe, any duplicates are dropped and a sample of 10 songs is drawn for the playlist.

Here I am creating a similarity matrix, deleting songs that both dataframes contain first. (common_songs)

In [17]:
unique_top_tracks_m = top_tracks_m[~top_tracks_m["track_id"].isin(common_songs["track_id"])]
unique_top_tracks_m.reset_index(drop = True,inplace = True)
unique_top_tracks_m

track_name                 album                track_id  \
0            Lambo Lambo           KitschKrieg  7oqvRZNv4dUV8CgQWtIAMe   
1             By My Side       Ancient Shadows  0eSWvkrCADGx0409mebsP4   
2            DEUTSCHLAND             RAMMSTEIN  2bPGTMB5sFfFYQ2YvSmup0   
3                 Zunder           Bittersweet  10Sp3ZHJUkSoYCN9NZO7QL   
4        Bande organisée       Bande organisée  205HNJ73cgpC0LAOnuQiWT   
..                   ...                   ...                     ...   
126            Hell Yeah           Light It Up  0GRu2odJjampqZpyubbFU6   
127            Like A G6             Like a G6  5S1yeuswEPofFOCkMFLNLc   
128             Halftime               Wolke 7  1N9ZAbqVw5o0m7wccSgRIt   
129              Push It  Wisconsin Death Trip  43WFwjiWFHc8ZryT1Tz1aY   
130  Talk Shit, Get Shot          Manslaughter  2rsho1vI3L5o65QOjM5Mwx   

                artist  duration  popularity  danceability  energy  key  \
0          KitschKrieg    214991          59         0.876   0.400    5   
1               Ecepta    193804          38         0.587   0.529    7   
2            Rammstein    322339          74         0.521   0.895    7   
3        Marek Hemmann    306976          48         0.811   0.639    9   
4                 Kofs    356346          79         0.901   0.939    6   
..                 ...       ...         ...           ...     ...  ...   
126         Rev Theory    247613           0         0.361   0.983    3   
127  Far East Movement    218346           0         0.629   0.869    7   
128               Gzuz    181060           1         0.533   0.783    2   
129           Static-X    154906          36         0.557   0.977    0   
130         Body Count    227147          39         0.455   0.993    3   

     loudness  mode  speechiness  acousticness  instrumentalness  liveness  \
0      -9.748     0       0.1290      0.494000          0.000009    0.1000   
1     -13.713     0       0.0376      0.285000          0.729000    0.1470   
2      -5.242     1       0.0442      0.000055          0.349000    0.0985   
3      -9.956     1       0.0515      0.011800          0.893000    0.0850   
4      -2.762     1       0.2740      0.117000          0.000000    0.0643   
..        ...   ...          ...           ...               ...       ...   
126    -3.373     0       0.0931      0.000129          0.000004    0.0228   
127    -7.013     0       0.3140      0.006570          0.000000    0.1910   
128    -6.490     1       0.3250      0.181000          0.000000    0.1010   
129    -3.965     1       0.0493      0.000393          0.000004    0.2910   
130    -1.952     0       0.2930      0.002350          0.000000    0.1490   

     valence    tempo  
0      0.484  144.938  
1      0.175  138.047  
2      0.237  120.117  
3      0.277  126.030  
4      0.805  142.948  
..       ...      ...  
126    0.385  167.951  
127    0.715  125.024  
128    0.471  129.844  
129    0.504  149.879  
130    0.384  175.004  

[131 rows x 17 columns]

In [18]:
unique_top_tracks_t = top_tracks_t[~top_tracks_t["track_id"].isin(common_songs["track_id"])]
unique_top_tracks_t.reset_index(drop = True,inplace = True)
unique_top_tracks_t

track_name                               album  \
0                        Time for Space                             Baralku   
1    Take Some Time - Emancipator Remix  Take Some Time (Emancipator Remix)   
2                                  Bati                     Chelina, Vol. 1   
3                        Intro (Megbia)                     Chelina, Vol. 1   
4                             Greenland            Safe In the Steep Cliffs   
..                                  ...                                 ...   
129               The Harder the Ground                                   Ø   
130             Smek - Rey&Kjavik Remix             Smek (Rey&Kjavik Remix)   
131                           Baby Girl                           Baby Girl   
132               Will We Ever Carry On                             Chances   
133                            Uma Vida                            Uma Vida   

                   track_id       artist  duration  popularity  danceability  \
0    3dJXvBddoH1AGLpKvmbYDA  Emancipator    433613          56         0.610   
1    28M2ugvRSIa4MIKmiiwNao    Wilderado    290428          41         0.545   
2    03xV05Oll19y59x8GDkWVL      Chelina    260713          10         0.575   
3    0aLjvBLCCpNNoBwGqNY6Gn      Chelina     53034           5         0.554   
4    2SPTGg9SC5MT1FwNX4IYfx  Emancipator    191066          57         0.593   
..                      ...          ...       ...         ...           ...   
129  5hb7SiqKHDlMPzs9HEholc        Gløde    243979          13         0.533   
130  1PyfWp1rg8omX2DOtg65pR         Ÿuma    366006          53         0.892   
131  6DNTJAvdyDwkmiPm2Wl8At   Bryce Vine    145371          55         0.860   
132  2AMAh0IXu1Wwwdw3Jzojqt  Luke Marzec    254549          17         0.638   
133  46qMKSFypTgW2Yk5dRyWgN      Gerra G    399183          16         0.757   

     energy  key  loudness  mode  speechiness  acousticness  instrumentalness  \
0    0.3130    5   -13.138     0       0.0323        0.3900          0.840000   
1    0.5960    5    -8.216     1       0.0289        0.0249          0.025200   
2    0.3610    2    -8.074     0       0.0291        0.5840          0.010300   
3    0.0872    9   -17.823     0       0.1510        0.9090          0.000000   
4    0.3980    4   -13.301     0       0.0295        0.0273          0.827000   
..      ...  ...       ...   ...          ...           ...               ...   
129  0.2400    5   -14.995     1       0.0370        0.9230          0.000239   
130  0.2810    2   -13.880     1       0.0435        0.8560          0.878000   
131  0.5920    2    -8.606     1       0.0861        0.1660          0.000609   
132  0.2930    0   -15.712     0       0.0337        0.5550          0.505000   
133  0.8290    2    -7.263     1       0.0454        0.1860          0.631000   

     liveness  valence    tempo  
0      0.0899   0.0870   80.009  
1      0.2190   0.1820   95.002  
2      0.1170   0.4300  148.154  
3      0.1160   0.3280   68.149  
4      0.2310   0.0989   98.728  
..        ...      ...      ...  
129    0.0877   0.0743  126.007  
130    0.1950   0.0433   99.984  
131    0.2760   0.6210  107.016  
132    0.1110   0.2260   81.802  
133    0.1040   0.6140   97.987  

[134 rows x 17 columns]

In [19]:
similarity_top_songs = create_similarity_score(unique_top_tracks_m,unique_top_tracks_t)

In [20]:
similarity_top_songs

array([[ 0.00246244, -0.73826128,  0.35275185, ..., -0.15783126,
         0.32730655, -0.46802349],
       [ 0.69407693, -0.31286898, -0.05266639, ..., -0.61516896,
         0.7241846 , -0.29211147],
       [ 0.42584099,  0.5070319 , -0.12468504, ..., -0.36538411,
         0.0779511 ,  0.47104614],
       ...,
       [-0.38903268, -0.26586009,  0.11660654, ..., -0.20057975,
        -0.30877886, -0.05457458],
       [-0.55930898,  0.44139615,  0.28426055, ...,  0.35761975,
        -0.68306416,  0.493662  ],
       [-0.57500067,  0.18209233,  0.28468891, ..., -0.0342616 ,
        -0.81988793,  0.19045593]])

Creating a list of tuples containing the indeces for both songs as well as the similarity score.

In [21]:
max_n_scores = [(i,np.argmax(x),x[np.argmax(x)]) for i,x in enumerate(similarity_top_songs)]
max_n_scores

[(0, 42, 0.9124961743889184),
 (1, 124, 0.9094278322775948),
 (2, 119, 0.7791874952042619),
 (3, 66, 0.9054499946753933),
 (4, 113, 0.8561162515101026),
 (5, 55, 0.8512760488540422),
 (6, 44, 0.9314149567120572),
 (7, 105, 0.9013989196086852),
 (8, 38, 0.7391856681964181),
 (9, 3, 0.7053260881584944),
 (10, 118, 0.8136696515203443),
 (11, 10, 0.9333967774914899),
 (12, 131, 0.7826843319439812),
 (13, 32, 0.7728838619396077),
 (14, 69, 0.8812387238367473),
 (15, 81, 0.779141363403892),
 (16, 101, 0.822876897994827),
 (17, 91, 0.8971764220497673),
 (18, 66, 0.8916073063101455),
 (19, 124, 0.9018786092021321),
 (20, 107, 0.9023017741771021),
 (21, 100, 0.9640395864441592),
 (22, 1, 0.8889369017104372),
 (23, 38, 0.7101835241708665),
 (24, 30, 0.740927735803688),
 (25, 107, 0.8665741787095874),
 (26, 74, 0.7798559023302319),
 (27, 46, 0.7664194631547017),
 (28, 10, 0.8982786919690497),
 (29, 102, 0.8668800750402578),
 (30, 124, 0.957576231037294),
 (31, 81, 0.7505792289398938),
 (32, 61, 0

In [22]:
from operator import itemgetter
from heapq import nlargest
nlargest(30,max_n_scores,key=itemgetter(2))

[(21, 100, 0.9640395864441592),
 (30, 124, 0.957576231037294),
 (82, 23, 0.9536280513031131),
 (123, 18, 0.944921194677877),
 (87, 43, 0.9443746258098373),
 (38, 105, 0.9346037773901044),
 (11, 10, 0.9333967774914899),
 (96, 121, 0.9326737211369086),
 (6, 44, 0.9314149567120572),
 (104, 16, 0.9287322242348719),
 (111, 10, 0.9253705955960095),
 (67, 108, 0.9247256489537813),
 (93, 16, 0.921148619965111),
 (34, 61, 0.9181729266168114),
 (109, 107, 0.916349603737838),
 (0, 42, 0.9124961743889184),
 (1, 124, 0.9094278322775948),
 (126, 106, 0.9087628594217892),
 (3, 66, 0.9054499946753933),
 (20, 107, 0.9023017741771021),
 (19, 124, 0.9018786092021321),
 (7, 105, 0.9013989196086852),
 (42, 58, 0.9012143244956449),
 (28, 10, 0.8982786919690497),
 (41, 126, 0.8976544159880666),
 (17, 91, 0.8971764220497673),
 (84, 94, 0.8943283062283301),
 (56, 106, 0.8922739791868055),
 (18, 66, 0.8916073063101455),
 (22, 1, 0.8889369017104372)]

Extracting the track pairs with the 30 highest similarity scores for each user.

In [23]:
idx_simtracks_m = [i[0] for i in  nlargest(30,max_n_scores,key=itemgetter(2))]
idx_simtracks_t = [i[1] for i in  nlargest(30,max_n_scores,key=itemgetter(2))]

In [24]:
idx_simtracks_m

[21,
 30,
 82,
 123,
 87,
 38,
 11,
 96,
 6,
 104,
 111,
 67,
 93,
 34,
 109,
 0,
 1,
 126,
 3,
 20,
 19,
 7,
 42,
 28,
 41,
 17,
 84,
 56,
 18,
 22]

In [25]:
idx_simtracks_t

[100,
 124,
 23,
 18,
 43,
 105,
 10,
 121,
 44,
 16,
 10,
 108,
 16,
 61,
 107,
 42,
 124,
 106,
 66,
 107,
 124,
 105,
 58,
 10,
 126,
 91,
 94,
 106,
 66,
 1]

In [26]:
sim_top_tracks_m = unique_top_tracks_m.loc[idx_simtracks_m]
sim_top_tracks_t = unique_top_tracks_t.loc[idx_simtracks_t]

Creating the dataframe with the most similar top tracks.

In [27]:
similar_top_tracks = pd.concat([sim_top_tracks_m,sim_top_tracks_t])
similar_top_tracks.drop_duplicates(inplace = True)
similar_top_tracks = similar_top_tracks[~similar_top_tracks["track_id"].isin(last_week_duo["track_id"])]
similar_top_tracks.reset_index(drop = True,inplace = True)
similar_top_tracks

track_name  \
0                                          Дни и ночи   
1                                  The Flying Octopus   
2                                             Phoenix   
3                                      WAIDMANNS HEIL   
4                                                Only   
5                                            Handsome   
6                                               Siren   
7   Points of Authority / 99 Problems / One Step C...   
8                                             Zuhause   
9                             Movin' Bass - GTA Remix   
10                                         Until Dawn   
11                            Movin' Bass - GTA Remix   
12                                         Gold Cobra   
13                                    That Other Girl   
14                                     NEW MAGIC WAND   
15                                         By My Side   
16                                          Hell Yeah   
17                                             Zunder   
18                                           Oh Junge   
19                                        Rolls Royce   
20                                    Vanishing Point   
21                                          Evolution   
22                                 FEIND (feat. Azad)   
23                                          Hurricane   
24                                       Blooming Era   
25                                              Sonne   
26                                     Wonderful Life   
27                                  Echo of the Woods   
28                              Wenn der Winter kommt   
29                                        Delmenhorst   
30                                   Phía Sau Đôi Mắt   
31                                     Painting Greys   
32                                              Insan   
33                                       Cool Me Down   
34                       Colour - Thousand Island Mix   
35                                             Endesu   
36                                         Bella ciao   
37                                              Kylie   
38          Comets (feat. Natalia Doco) - HUGEL Remix   
39                                    Galopa, Murieta   
40                                 When We Were Young   
41                             Scheune - Original Mix   
42                                  Keiner hat gesagt   
43                                        Kleines Rad   
44                           Moon Over Bourbon Street   
45                                Was du Liebe nennst   

                                            album                track_id  \
0                                      Дни и ночи  3A88HuZw5yiZOAcon4RkZk   
1                                         Charlie  6bv9gQNLWUUiuzRunuY4lJ   
2                                         Phoenix  6zAiRKvAMlXHxEtyO4yxIO   
3         LIEBE IST FÜR ALLE DA (SPECIAL EDITION)  6ey8jr96hlSORKyeJk9d8d   
4   The Pinkprint (International Deluxe Explicit)  1UZ25gykR30Oewh3dBRtVZ   
5                               Heavy Is The Head  766N0mjf1KzAAYsA1eAJaN   
6                                           Siren  5umBsHgpB2WsRA9ccQZGdz   
7               Collision Course (Deluxe Version)  65eohvrL4ttjA7EfFkQOhX   
8                                             Nie  5vC4M4GjYLkgDaUwQcL7WA   
9                         Movin' Bass (GTA Remix)  2BnVVl5NCV2o7XlZo1CLEm   
10                                     Until Dawn  1R2rmmUufa1tdiD9j9Qs3x   
11                        Movin' Bass (GTA Remix)  7yzwszgJUDrllWnG9cQkxQ   
12                            Gold Cobra (Deluxe)  7AzWF3NULrfkluXrN7N8fK   
13                              The Suspended Kid  04Mr5jeXzPFnEQUPtPzRVP   
14                                           IGOR  0fv2KH6hac06J86hBUTcSf   
15                                Ancient Shadows  0eSWvkrCADGx0409mebsP4   
16                                    Light It Up  0GR

In [28]:
similar_top_tracks.sample(10)

track_name  \
30                                   Phía Sau Đôi Mắt   
8                                             Zuhause   
5                                            Handsome   
26                                     Wonderful Life   
14                                     NEW MAGIC WAND   
33                                       Cool Me Down   
11                            Movin' Bass - GTA Remix   
36                                         Bella ciao   
27                                  Echo of the Woods   
7   Points of Authority / 99 Problems / One Step C...   

                                album                track_id  \
30                   Phía Sau Đôi Mắt  3DJZM2sfbUleDaCLRXWSco   
8                                 Nie  5vC4M4GjYLkgDaUwQcL7WA   
5                   Heavy Is The Head  766N0mjf1KzAAYsA1eAJaN   
26                     Wonderful Life  0cqLFzFYsQurUI3McUZkLg   
14                               IGOR  0fv2KH6hac06J86hBUTcSf   
33                       Cool Me Down  0oDpBQNBCa3VCR3bmn7953   
11            Movin' Bass (GTA Remix)  7yzwszgJUDrllWnG9cQkxQ   
36                         Bella ciao  51frPF1JFad9MBlgI3dg1J   
27                  Echo of the Woods  6d8mjJCtnlFtgGv3erEs7c   
7   Collision Course (Deluxe Version)  65eohvrL4ttjA7EfFkQOhX   

                artist  duration  popularity  danceability  energy  key  \
30       Anh bạn Thành    174206          36         0.545   0.398    7   
8        Fynn Kliemann    197887          65         0.656   0.448    9   
5              Stormzy    152733          56         0.762   0.559    1   
26         Katie Melua    247152          61         0.904   0.341    6   
14  Tyler, The Creator    195320          70         0.621   0.730    5   
33            Margaret    179211          43         0.647   0.716    2   
11           Rick Ross    253773          52         0.698   0.911   11   
36         Mike Singer    184000          42         0.764   0.645   11   
27               Giolì    198260          37         0.683   0.314    5   
7                JAY-Z    295826          58         0.547   0.951    1   

    loudness  mode  speechiness  acousticness  instrumentalness  liveness  \
30   -13.675     0       0.3170      0.905000          0.000242    0.1120   
8     -8.676     0       0.0442      0.860000          0.000005    0.1030   
5     -6.854     1       0.1530      0.173000          0.000000    0.1270   
26   -12.088     0       0.0417      0.792000          0.021000    0.0977   
14    -5.414     0       0.1070      0.096700          0.000131    0.6730   
33    -4.590     1       0.0483      0.004370          0.000983    0.0997   
11    -2.570     0       0.2170      0.000757          0.000103    0.2810   
36    -4.652     1       0.0501      0.191000          0.000082    0.1590   
27   -15.354     0       0.0385      0.768000          0.892000    0.1550   
7     -4.079     0       0.1510      0.002720          0.000011    0.0862   

    valence    tempo  
30   0.1560  145.513  
8    0.4030   95.626  
5    0.2820  139.955  
26   0.4470  105.179  
14   0.4640  139.566  
33   0.4560   90.995  
11   0.5640  160.031  
36   0.6830  144.019  
27   0.0409  115.024  
7    0.5100   95.031

In [30]:
new_playlist_df = new_playlist_df.append(similar_top_tracks.sample(10))
new_playlist_df

track_name  \
3                              Huldra - Other Version   
0                                          Дни и ночи   
2                                             Phoenix   
17                                             Zunder   
15                                         By My Side   
28                              Wenn der Winter kommt   
19                                        Rolls Royce   
7   Points of Authority / 99 Problems / One Step C...   
3                                      WAIDMANNS HEIL   
5                                            Handsome   
27                                  Echo of the Woods   

                                      album                track_id  \
3                                    Huldra  569gNjph2g07MmjtMm6vKm   
0                                Дни и ночи  3A88HuZw5yiZOAcon4RkZk   
2                                   Phoenix  6zAiRKvAMlXHxEtyO4yxIO   
17                              Bittersweet  10Sp3ZHJUkSoYCN9NZO7QL   
15                          Ancient Shadows  0eSWvkrCADGx0409mebsP4   
28                     Mittelpunkt der Welt  6aj2zW1HmhZNQoeGR1avif   
19                              Rolls Royce  22XFe65IH0P2RY0uowBqdI   
7         Collision Course (Deluxe Version)  65eohvrL4ttjA7EfFkQOhX   
3   LIEBE IST FÜR ALLE DA (SPECIAL EDITION)  6ey8jr96hlSORKyeJk9d8d   
5                         Heavy Is The Head  766N0mjf1KzAAYsA1eAJaN   
27                        Echo of the Woods  6d8mjJCtnlFtgGv3erEs7c   

               artist  duration  popularity  danceability  energy  key  \
3               Gidge    490000          38         0.772   0.514   10   
0              Geegun    186000          48         0.917   0.526    2   
2   League of Legends    197633          75         0.420   0.723   10   
17      Marek Hemmann    306976          48         0.811   0.639    9   
15             Ecepta    193804          38         0.587   0.529    7   
28   Element Of Crime    274306          29         0.332   0.454    4   
19             Geegun    143437          68         0.872   0.423    1   
7               JAY-Z    295826          58         0.547   0.951    1   
3           Rammstein    212800           0         0.578   0.969    9   
5             Stormzy    152733          56         0.762   0.559    1   
27              Giolì    198260          37         0.683   0.314    5   

    loudness  mode  speechiness  acousticness  instrumentalness  liveness  \
3    -11.961     0       0.0754       0.14800          0.906000    0.0934   
0     -8.150     0       0.1250       0.29800          0.000058    0.1080   
2     -6.860     0       0.1750       0.06380          0.000012    0.1230   
17    -9.956     1       0.0515       0.01180          0.893000    0.0850   
15   -13.713     0       0.0376       0.28500          0.729000    0.1470   
28    -8.373     1       0.0443       0.27800          0.000000    0.1100   
19    -7.337     1       0.2130       0.14200          0.000010    0.1040   
7     -4.079     0       0.1510       0.00272          0.000011    0.0862   
3     -3.894     1       0.0317       0.00113          0.023400    0.0579   
5     -6.854     1       0.1530       0.17300          0.000000    0.1270   
27   -15.354     0       0.0385       0.76800          0.892000    0.1550   

    valence    tempo  
3    0.1030  118.015  
0    0.5140  119.985  
2    0.2880  167.806  
17   0.2770  126.030  
15   0.1750  138.047  
28   0.2440  193.468  
19   0.3730   95.980  
7    0.5100   95.031  
3    0.6840  104.999  
5    0.2820  139.955  
27   0.0409  115.024

### Sampling from each users top tracks which are from an artists both users like (one of their top artists)

The next step filters the top tracks based on common artists. Every track by an artist that is a top artist of **both** users is considered for this approach.



In [31]:
filtered_top_m = top_tracks_m[top_tracks_m["artist"].isin(common_artists["name"]) 
                              & ~top_tracks_m["track_id"].isin(last_week_duo["track_id"])]
filtered_top_m

track_name        album                track_id        artist  \
9                 Oh Junge  KitschKrieg  0JqbpesudPE6j901fBEzo2   KitschKrieg   
15               Zim Zimma    Evolution  17nPeSliosCi427f0lUb75  Joyner Lucas   
37                    ADHD         ADHD  4X4v3KtkUXwXvDBw5KS9cp  Joyner Lucas   
46             Fall Slowly  Fall Slowly  01WOwxkxOw2FqNIHkraxcN  Joyner Lucas   
47               Evolution    Evolution  2VopDw2GlF3uwD1kihHmTT  Joyner Lucas   
51             ALLES HELAL        NACHT  2KAbQ3PsETrr86R39pru7k          ELIF   
57                   FEUER        NACHT  0Se4w42WIJiTgld4SYbv8S          ELIF   
61        ALL OF MY DREAMS        NACHT  41R2FrKYRgLHntLBpU4NXE          ELIF   
63                 SCHWARZ        NACHT  7vswtCdKBzC5XN9ojwh8u0          ELIF   
64                  ALASKA        NACHT  2YnYp5f38UP6fvf7q2FnPm          ELIF   
65         ABER WO BIST DU        NACHT  2dGtLJysTSI9cbQ6TulL8V          ELIF   
69         EIN LETZTES MAL        NACHT  04EJyZSlhPPfFOo1NRn2vl          ELIF   
77                 NUR MIR        NACHT  1jzoAcvAE7vykG1sTI2mvt          ELIF   
79  AUGEN ZU (feat. Samra)        NACHT  6ikiAn9th3TUfS3bYS7gDX          ELIF   
80                 GITARRE        NACHT  3za0JhyT84HuHOj2iLY9h5          ELIF   
86   DU TUST MIR NICHT GUT        NACHT  2pmqIWXxhdvjJfRiwySIZl          ELIF   
90      FEIND (feat. Azad)        NACHT  4Upt2Q8OcUd85uo5x9QqWK          ELIF   

    duration  popularity  danceability  energy  key  loudness  mode  \
9     193946          57         0.878   0.515    1    -7.715     1   
15    239702          69         0.883   0.621   11    -6.063     0   
37    205872          70         0.563   0.780   10    -6.663     1   
46    208095          63         0.576   0.534    6    -9.637     0   
47    153250          60         0.687   0.819    9    -6.670     0   
51    175062          60         0.727   0.600    4    -6.186     0   
57    203386          45         0.769   0.604    8    -6.769     1   
61    154676          42         0.601   0.619    1    -5.815     0   
63    135184          45         0.742   0.700   10    -6.937     0   
64    217766          48         0.743   0.671    1    -5.594     0   
65    185274          55         0.780   0.592    9    -6.020     0   
69    144644          50         0.713   0.549    1    -6.168     1   
77    150136          46         0.737   0.535    0    -6.129     0   
79    174248          70         0.685   0.693    5    -6.741     0   
80    188484          42         0.810   0.743    5    -5.664     0   
86    207279          52         0.860   0.434    4    -7.951     0   
90    169299          52         0.781   0.770    1    -5.367     1   

    speechiness  acousticness  instrumentalness  liveness  valence    tempo  
9        0.1880       0.02920          0.000000    0.0835    0.269  156.000  
15       0.2120       0.08710          0.000000    0.4990    0.676  149.052  
37       0.0782       0.00525          0.000008    0.4180    0.317   83.913  
46       0.2940       0.07500          0.000000    0.1990    0.041   83.560  
47       0.4310       0.21800          0.000000    0.3920    0.568   81.185  
51       0.0376       0.16400          0.000017    0.0793    0.144   92.024  
57       0.0455       0.03900          0.000277    0.0738    0.227   97.996  
61       0.2420       0.05080          0.000001    0.0555    0.146  172.346  
63       0.1770       0.09490          0.000037    0.1040    0.465   75.985  
64       0.0589       0.05810          0.000000    0.1950    0.484   76.998  
65       0.0631       0.23000          0.000000    0.1840    0.705  148.057  
69       0.0645       0.04610          0.000041    0.1080    0.393   89.005  
77       0.0549       0.01970          0.000021    0.1160    0.635   83.046  
79       0.2170       0.12500          0.000002    0.2750    0.493  167.957  
80       0.0348       0.03260          0.000000    0.1250    0.272  125.979  
86       0.0677       0.63000 

In [32]:
filtered_top_t = top_tracks_t[top_tracks_t["artist"].isin(common_artists["name"])
                             & ~top_tracks_t["track_id"].isin(last_week_duo["track_id"])]
filtered_top_t

track_name    album                track_id  \
105                        Lass Sie Gehn  BAM BAM  4RYKr1R3tXrITqY1zWiTNi   
107  Lass Das Licht An (feat. Deichkind)  BAM BAM  0BJ5N5yfEPBAimh9LprQvc   
126        Aufstehn! (feat. CeeLo Green)    Next!  2KpvM181ZwtIWwIElZx4vI   
128          Sie Is Geladen (feat. Nura)  BAM BAM  4Qt7LInzG4ruqcIrowldhg   

    artist  duration  popularity  danceability  energy  key  loudness  mode  \
105  Seeed    192853          54         0.711   0.706    7    -4.543     1   
107  Seeed    188561          50         0.904   0.664    2    -4.637     1   
126  Seeed    230960          59         0.849   0.830    6    -3.907     1   
128  Seeed    218702          46         0.685   0.885    9    -4.248     0   

     speechiness  acousticness  instrumentalness  liveness  valence    tempo  
105        0.289        0.0776          0.000000     0.127    0.695  138.810  
107        0.288        0.0753          0.000000     0.151    0.720  143.746  
126        0.186        0.1120          0.000000     0.384    0.873  127.279  
128        0.211        0.0168          0.000023     0.179    0.629  144.050

#### Potential issues

This filtering approach often leads to lists of songs that only contain very few artists but a couple of songs by that artist (it's also due to spotify really noticing when you can't stop listening to an album..)
To not have too many songs by the same artist I will sample from the above dataframes.I am assigning weights to the rows depending on how often an artist occurs.

This approach worked reasonably well however still has some flaws (which might be partially driven by my girlfriends and my individual listening behavior)

In [33]:
from collections import Counter

Counter(filtered_top_m["artist"]) ,  Counter(filtered_top_t["artist"])

(Counter({'KitschKrieg': 1, 'Joyner Lucas': 4, 'ELIF': 12}),
 Counter({'Seeed': 4}))

In [34]:
weights_m = [1/len(filtered_top_m)/7 if Counter(filtered_top_m["artist"])[x] > 2 else 1/len(filtered_top_m) for x in filtered_top_m["artist"]] 

In [35]:
weights_t = [1/len(filtered_top_t)/7 if Counter(filtered_top_t["artist"])[x] > 2 else 1/len(filtered_top_m) for x in filtered_top_t["artist"]] 

I tried the sampling with and without weights 10 times. With weights artists that occur very often in the filtered dataframe are not too overrepresented just as planned. Without weights the sample sometimes contained just one or two artists which is not desired.

In [36]:
sample_n = (25-len(new_playlist_df))//2
if sample_n > 3: sample_n = 3
sample_n

3

In [37]:
filtered_top_m.sample(sample_n,weights = weights_m)

track_name        album                track_id        artist  duration  \
57      FEUER        NACHT  0Se4w42WIJiTgld4SYbv8S          ELIF    203386   
9    Oh Junge  KitschKrieg  0JqbpesudPE6j901fBEzo2   KitschKrieg    193946   
47  Evolution    Evolution  2VopDw2GlF3uwD1kihHmTT  Joyner Lucas    153250   

    popularity  danceability  energy  key  loudness  mode  speechiness  \
57          45         0.769   0.604    8    -6.769     1       0.0455   
9           57         0.878   0.515    1    -7.715     1       0.1880   
47          60         0.687   0.819    9    -6.670     0       0.4310   

    acousticness  instrumentalness  liveness  valence    tempo  
57        0.0390          0.000277    0.0738    0.227   97.996  
9         0.0292          0.000000    0.0835    0.269  156.000  
47        0.2180          0.000000    0.3920    0.568   81.185

In [38]:
filtered_top_t.sample(sample_n, weights= weights_t)

track_name    album                track_id artist  \
105                  Lass Sie Gehn  BAM BAM  4RYKr1R3tXrITqY1zWiTNi  Seeed   
128    Sie Is Geladen (feat. Nura)  BAM BAM  4Qt7LInzG4ruqcIrowldhg  Seeed   
126  Aufstehn! (feat. CeeLo Green)    Next!  2KpvM181ZwtIWwIElZx4vI  Seeed   

     duration  popularity  danceability  energy  key  loudness  mode  \
105    192853          54         0.711   0.706    7    -4.543     1   
128    218702          46         0.685   0.885    9    -4.248     0   
126    230960          59         0.849   0.830    6    -3.907     1   

     speechiness  acousticness  instrumentalness  liveness  valence    tempo  
105        0.289        0.0776          0.000000     0.127    0.695  138.810  
128        0.211        0.0168          0.000023     0.179    0.629  144.050  
126        0.186        0.1120          0.000000     0.384    0.873  127.279

In [39]:
new_playlist_df = new_playlist_df.append(filtered_top_m.sample(sample_n,weights = weights_m))
new_playlist_df = new_playlist_df.append(filtered_top_t.sample(sample_n,weights = weights_t))

In [40]:
new_playlist_df = new_playlist_df.drop_duplicates().reset_index(drop=True)
new_playlist_df

track_name  \
0                              Huldra - Other Version   
1                                          Дни и ночи   
2                                             Phoenix   
3                                              Zunder   
4                                          By My Side   
5                               Wenn der Winter kommt   
6                                         Rolls Royce   
7   Points of Authority / 99 Problems / One Step C...   
8                                      WAIDMANNS HEIL   
9                                            Handsome   
10                                  Echo of the Woods   
11                                    EIN LETZTES MAL   
12                                    ABER WO BIST DU   
13                                          Zim Zimma   
14                Lass Das Licht An (feat. Deichkind)   
15                                      Lass Sie Gehn   
16                      Aufstehn! (feat. CeeLo Green)   

                                      album                track_id  \
0                                    Huldra  569gNjph2g07MmjtMm6vKm   
1                                Дни и ночи  3A88HuZw5yiZOAcon4RkZk   
2                                   Phoenix  6zAiRKvAMlXHxEtyO4yxIO   
3                               Bittersweet  10Sp3ZHJUkSoYCN9NZO7QL   
4                           Ancient Shadows  0eSWvkrCADGx0409mebsP4   
5                      Mittelpunkt der Welt  6aj2zW1HmhZNQoeGR1avif   
6                               Rolls Royce  22XFe65IH0P2RY0uowBqdI   
7         Collision Course (Deluxe Version)  65eohvrL4ttjA7EfFkQOhX   
8   LIEBE IST FÜR ALLE DA (SPECIAL EDITION)  6ey8jr96hlSORKyeJk9d8d   
9                         Heavy Is The Head  766N0mjf1KzAAYsA1eAJaN   
10                        Echo of the Woods  6d8mjJCtnlFtgGv3erEs7c   
11                                    NACHT  04EJyZSlhPPfFOo1NRn2vl   
12                                    NACHT  2dGtLJysTSI9cbQ6TulL8V   
13                                Evolution  17nPeSliosCi427f0lUb75   
14                                  BAM BAM  0BJ5N5yfEPBAimh9LprQvc   
15                                  BAM BAM  4RYKr1R3tXrITqY1zWiTNi   
16                                    Next!  2KpvM181ZwtIWwIElZx4vI   

               artist  duration  popularity  danceability  energy  key  \
0               Gidge    490000          38         0.772   0.514   10   
1              Geegun    186000          48         0.917   0.526    2   
2   League of Legends    197633          75         0.420   0.723   10   
3       Marek Hemmann    306976          48         0.811   0.639    9   
4              Ecepta    193804          38         0.587   0.529    7   
5    Element Of Crime    274306          29         0.332   0.454    4   
6              Geegun    143437          68         0.872   0.423    1   
7               JAY-Z    295826          58         0.547   0.951    1   
8           Rammstein    212800           0         0.578   0.969    9   
9             Stormzy    152733          56         0.762   0.559    1   
10              Giolì    198260          37         0.683   0.314    5   
11               ELIF    144644          50         0.713   0.549    1   
12               ELIF    185274          55         0.780   0.592    9   
13       Joyner Lucas    239702          69         0.883   0.621   11   
14              Seeed    188561          50         0.904   0.664    2   
15              Seeed    192853          54         0.711   0.706    7   
16              Seeed    230960          59         0.849   0.830    6   

    loudness  mode  speechiness  acousticness  instrumentalness  liveness  \
0    -11.961     0       0.0754       0.14800          0.906000    0.0934   
1     -8.150     0       0.1250       0.29800          0.000058    0.1080   
2     -6.860     0       0.1750       0.06380          0.000012    0.1230   
3     -9.956     1       0.0515       0.01180          0.893000    0.0850   
4    -13.713     0 

### Sampling from saved tracks

I am aiming for around 25 known tracks (and 25 new ones through recommendations). To achieve this and to somehow account for the somewhat random nature of the previous steps I am filling the playlist with sampled saved tracks.

In [41]:
#sample the remaining 25-len(new_playlist_df) from saved_tracks
#first get audio_features
saved_tracks_m = append_audio_features(saved_tracks_m, sp_m)
saved_tracks_t = append_audio_features(saved_tracks_t,sp_t)

In [42]:
#filter again so artists are not already in new_playlist_df
filtered_saved_m = saved_tracks_m[~saved_tracks_m["artist"].isin(new_playlist_df["artist"])]
filtered_saved_m

track_name  \
0                               From America   
2                              The Beginning   
3                            Between Breaths   
4                              Berlin Nights   
8   Money In The Grave (Drake ft. Rick Ross)   
9                              Beifahrersitz   
10                                    Twingo   
13                            When I'm Small   
14            Risky Business - Mathame Remix   
15                                      Aura   
16                                  Blackout   
17                      Who - Single Version   
18                               Ghostkeeper   
19                                   Falling   
20               Five Minutes - Zimmer Remix   
21                              Five Minutes   
22    The Man With the Red Face - Video Edit   
23        Man With The Red Face - Radio Edit   
24                            Slap the Ghost   
25                         You.Won't.Do.Shit   
26                                Full Metal   
27                        Lost In The Static   
28           Bounce Back (feat. Jay Anthony)   
29                                    So Far   
30                    Idol (feat. Tech N9ne)   
32                                      YAYA   
33                                Light Rain   
34                                  Over Now   
35                          Grave - Stripped   
36                                     ERROR   
37                                    Killer   
38                                   BURNING   
39                                     Alarm   
40                        Stargazing - Remix   
41                                    Omertà   
42       Generation YouPorn - Agia Napa Edit   
43                              C'est La Vie   
44                                    Dreams   
45                                Fire in Me   

                                              album                track_id  \
0                                     Play Me Again  5BMg8D9Wl4yvPqzTq7rWRC   
2                                     The Beginning  4njhWDhTAjhReWtYkiMH9t   
3                                         Stateless  4DJGTc1OsgqFsUGI6W8Mtx   
4                                     Berlin Nights  6gG1R1bFdJeNc2ERAwXxCb   
8                        The Best In The World Pack  5ry2OE6R2zPQFDO85XkgRb   
9                                     Beifahrersitz  01qOl2pM8emx1sxdBQc05g   
10                                              POP  6vFEvkXOnOTUMgPHxogIRK   
13                                    Eyelid Movies  3498wF96LsgVgMkGmJzJOC   
14                   Risky Business (Mathame Remix)  5R4hprpCcdgKz1DsPoh9p2   
15                         Aura (Exclusive Version)  6sDb7wNlVXQGqnhEHiNt8B   
16                                      EROS (2018)  7s2s2QBaO8cRhyBssmjI1q   
17                                              Who  0CjBORMsmiQNe3vPDcNIvk   
18                                      Ghostkeeper  13EMVBU1hcy5yJdX1Xhf9a   
19                                          Falling  4WA4jmG98kErhzC9oNf8Sx   
20                           Five Minutes (Remixes)  4X864m4tHKfMyfPUzJzqqd   
21                                     Five Minutes  3pJnvBbwIvuwebHHTYzoR3   
22                           Unreasonable Behaviour  3FwsjsakjJkeLotPOGIr7s   
23                            Man With The Red Face  5FDn1oWUD5BXL2HCfmCk2F   
24                         Slap the Ghost / Stinger  4RRNgDu0Mas7w3DUZm62Jk   
25                                        SWAN DIVE  61wOkYtKm5AwDbIhl05W0C   
26                                       Full Metal  3oMueNQTlVgOdIfFP9ENz9   
27                               Lost In The Static  6iGXUJW99KAwbRiwxIjKPU   
28                  Bounce Back (feat. Jay Anthony)  6zsPaeGXa4UxM3EBzlFyjN   
29  Broadchurch (Music From The Original TV Series)  6oVhL0lLUMswqSV3VcKwJO   
30                           Idol (feat. Tech N9ne)  3bLU7rl0xkgzcLPYHO3HuK   
32                                             YAYA  6EfBMJ

In [43]:
filtered_saved_t = saved_tracks_t[~saved_tracks_t["artist"].isin(new_playlist_df["artist"])]
filtered_saved_t

track_name  \
0                                        Крыльями   
1             Last Night in Sant Celoni - 12" Mix   
2                                        Medellín   
3                                            Bati   
4                                         Sai Bai   
5                                           Пожар   
6                                           Insan   
7         The World Retreats - Marino Canal Remix   
8                                       Opa Gäärd   
9                                           sugar   
10                                 A Child's Tale   
11                                     Love Again   
12             Take Some Time - Emancipator Remix   
13                                 Time for Space   
14                          Shuggie - Bonus Track   
15                                     Borboletas   
16    When The Interlude Sounds Like Falling Rain   
17                               Phía Sau Đôi Mắt   
18  Suite No. 2 for Jazz Orchestra: VII. Waltz II   
19                                         Carmen   
20                                        Zukunft   
21                                    Bye Felicia   
22                                 I Had A Vision   
23                                 Lavish Lullaby   
24           Silver Tongue Devil (feat. Shenseea)   
25               Mystery Lady (feat. Don Toliver)   
26                                That Other Girl   
27                                         Weiter   
28                                   The Long Run   
29                                   Love Tonight   
30               Lily Moon - Isaac Chambers Remix   
31                     На своём вайбе (feat. Гуф)   
32                                          Мадам   
33                                    Ночной рейс   
34                           Один день в сентябре   
35                    С любимыми не расставайтесь   
36                              Ты - моя нежность   
37                               Забываю обещания   
38                                         Джадуа   
39                                      Не модные   
40                                        Не могу   
41                                 The Broken Man   
42           Shroom Room - Iorie Pathfinder Remix   
43                                   The Aviating   
44                                   Mind Doodles   
45                               Комната скомкана   
46                                      Колхозник   
47                         Если ты меня не любишь   
48                                        Fragile   
49                                 Песня без слов   

                                                album                track_id  \
0                                            Источник  1wAkp6u1biHGKX2q5IwBhf   
1         Last Night in Sant Celoni (feat. Jaz James)  6zAqmi9YkgnMxbqLcf7hDv   
2                                #EnAttendantMachakil  4Nmrb5DL6lN5kkb7LzX8yE   
3                                     Chelina, Vol. 1  03xV05Oll19y59x8GDkWVL   
4                                     Chelina, Vol. 1  3vAJlGxRNNYYMBJdnhxynu   
5                                          Расстояние  772bNQ8WjMAXwiBywGcHb7   
6                                               Insan  01ZAEc3eyGGJLLZuAgdSMD   
7             The World Retreats (Marino Canal Remix)  5dyVrDue2N9ArxW34JyeO4   
8                                          A Long Way  1eb0mORiTlz0OLkH0NPb9Z   
9                                        next chapter  5IjIbGO7lih9CVDBFLCtTT   
10                                          Phantasma  2JNa5xzODo5tiHDIvLPpGt   
11                                         Love Again  7pf01odICRjg7yGgDsnkAP   
12                 Take Some Time (Emancipator Remix)  28M2ugvRSIa4MIKmiiwNao   
13                                            Baralku  3dJXvBddoH1AGLpKvmbYDA   
14                                                Her  24V1nw6aGxnIVuWHEGH4e1   
15                                     

In [44]:
sample_n = (25-len(new_playlist_df))//2
sample_n

4

In [45]:
new_playlist_df = pd.concat([new_playlist_df,filtered_saved_m.sample(sample_n),filtered_saved_t.sample(sample_n)])
new_playlist_df.reset_index(drop = True, inplace= True)

In [46]:
new_playlist_df

track_name  \
0                              Huldra - Other Version   
1                                          Дни и ночи   
2                                             Phoenix   
3                                              Zunder   
4                                          By My Side   
5                               Wenn der Winter kommt   
6                                         Rolls Royce   
7   Points of Authority / 99 Problems / One Step C...   
8                                      WAIDMANNS HEIL   
9                                            Handsome   
10                                  Echo of the Woods   
11                                    EIN LETZTES MAL   
12                                    ABER WO BIST DU   
13                                          Zim Zimma   
14                Lass Das Licht An (feat. Deichkind)   
15                                      Lass Sie Gehn   
16                      Aufstehn! (feat. CeeLo Green)   
17                     Risky Business - Mathame Remix   
18                                      Berlin Nights   
19                                 Stargazing - Remix   
20                                      The Beginning   
21        When The Interlude Sounds Like Falling Rain   
22                                       Mind Doodles   
23                                            Fragile   
24                   Mystery Lady (feat. Don Toliver)   

                                      album                track_id  \
0                                    Huldra  569gNjph2g07MmjtMm6vKm   
1                                Дни и ночи  3A88HuZw5yiZOAcon4RkZk   
2                                   Phoenix  6zAiRKvAMlXHxEtyO4yxIO   
3                               Bittersweet  10Sp3ZHJUkSoYCN9NZO7QL   
4                           Ancient Shadows  0eSWvkrCADGx0409mebsP4   
5                      Mittelpunkt der Welt  6aj2zW1HmhZNQoeGR1avif   
6                               Rolls Royce  22XFe65IH0P2RY0uowBqdI   
7         Collision Course (Deluxe Version)  65eohvrL4ttjA7EfFkQOhX   
8   LIEBE IST FÜR ALLE DA (SPECIAL EDITION)  6ey8jr96hlSORKyeJk9d8d   
9                         Heavy Is The Head  766N0mjf1KzAAYsA1eAJaN   
10                        Echo of the Woods  6d8mjJCtnlFtgGv3erEs7c   
11                                    NACHT  04EJyZSlhPPfFOo1NRn2vl   
12                                    NACHT  2dGtLJysTSI9cbQ6TulL8V   
13                                Evolution  17nPeSliosCi427f0lUb75   
14                                  BAM BAM  0BJ5N5yfEPBAimh9LprQvc   
15                                  BAM BAM  4RYKr1R3tXrITqY1zWiTNi   
16                                    Next!  2KpvM181ZwtIWwIElZx4vI   
17           Risky Business (Mathame Remix)  5R4hprpCcdgKz1DsPoh9p2   
18                            Berlin Nights  6gG1R1bFdJeNc2ERAwXxCb   
19                       Stargazing (Remix)  1HK6TWeHG9q5upt2WcI0sw   
20                            The Beginning  4njhWDhTAjhReWtYkiMH9t   
21                     What Are You Into EP  5lWVqc9kPplvlDmLtFls02   
22                             Mind Doodles  4ZVqqf3Eo4uXSeFUrYD5lw   
23                                  Fragile  59g3Adj1Vdhja52OrhURdE   
24         Mystery Lady (feat. Don Toliver)  2Ee9amLUslOwgoJWZEpSSD   

               artist  duration  popularity  danceability  energy  key  \
0               Gidge    490000          38         0.772   0.514   10   
1              Geegun    186000          48         0.917   0.526    2   
2   League of Legends    197633          75         0.420   0.723   10   
3       Marek Hemmann    306976          48         0.811   0.639    9   
4              Ecepta    193804          38         0.587   0.529    7   
5    Element Of Crime    274306          29         0.332   0.454    4   
6              Geegun    143437          68         0.872   0.423    1   
7               JAY-Z    295826          58         0.547   0.951    1   
8           Rammstein    212800           0    

### Adding new tracks from Spotify recommendations

In this last step I add new tracks to fill up the other half of the playlist.

I **don't** want to simply add songs spotify recommends based on the songs, which are already in the playlist.

Therefore getting Spotify recommendations is only the first step. I am retrieving multiple songs recommendations per song, which are then filtered again based on similarity scoring.

In [47]:
seed_tracks = new_playlist_df["track_id"].tolist()
#seed_artists = artists_m["name"].tolist() + artists_t["name"].tolist()

In [48]:
len(seed_tracks)

25

Unfortunately **the Spotify API does not accept 25 seed tracks for a recommendation query**, I therefore am splitting up the process into "packages" of 5 seed tracks, retrieving 25 tracks per "package"

In [49]:
seed_tracks[:5], seed_tracks[5:10], seed_tracks[10:15]

(['569gNjph2g07MmjtMm6vKm',
  '3A88HuZw5yiZOAcon4RkZk',
  '6zAiRKvAMlXHxEtyO4yxIO',
  '10Sp3ZHJUkSoYCN9NZO7QL',
  '0eSWvkrCADGx0409mebsP4'],
 ['6aj2zW1HmhZNQoeGR1avif',
  '22XFe65IH0P2RY0uowBqdI',
  '65eohvrL4ttjA7EfFkQOhX',
  '6ey8jr96hlSORKyeJk9d8d',
  '766N0mjf1KzAAYsA1eAJaN'],
 ['6d8mjJCtnlFtgGv3erEs7c',
  '04EJyZSlhPPfFOo1NRn2vl',
  '2dGtLJysTSI9cbQ6TulL8V',
  '17nPeSliosCi427f0lUb75',
  '0BJ5N5yfEPBAimh9LprQvc'])

In [50]:
recomms = sp_m.recommendations(seed_tracks = seed_tracks[:5],limit = 25)

In [51]:
append_audio_features(create_df_recommendations(recomms),sp_m)

track_name  \
0               Warschauer Strasse - Original Version   
1                                            Sine Ira   
2                                            Norrland   
3                                           Принцесса   
4                   Spree Ahoi (feat. Steven Coulter)   
5                                             Hate Me   
6                                               Devil   
7                                                 Nil   
8                                Fireflies - Original   
9                                    Миллион алых роз   
10                                     For Seoul Pt I   
11                                               Lyre   
12                                             Nevada   
13                                  We Want Your Soul   
14                                              Bloom   
15                                           Импульсы   
16  Goodbye (feat. Soap&Skin) [Theme from Dark, A ...   
17                                           Jetlands   
18                                           Nahimana   
19                                           Sandburg   
20                                             Voices   
21                            Do Not Do Me (Like Dis)   
22                                              Hades   
23                                   Bilder mit Katze   
24                           Your Lips - Original Mix   

                                                album                track_id  \
0                                    Großstadtmärchen  54Ouh41CWAFULtChxDc2P3   
1                                 The Dawn: Chapter 1  4N4cETijYD18C7bSJl8Vxy   
2                                        Autumn Bells  7kdaeyC4sTR8TEsskWIjbQ   
3                                                 313  6d4YTLAXAMUC8vxsSyBkmt   
4                   Spree Ahoi (feat. Steven Coulter)  6wvp4Mh9QrPAcs6sWiTj4x   
5                                             Hate Me  5FgdPkPCffPktT5qnWls8Y   
6                                               Devil  4BVKDnYVUxbfBYJIip2RBp   
7                                                Mare  5G3ZKjCHie2Ikr3I4QCQGt   
8                                           Fireflies  60NRvqEVjGSbocmylxfhDf   
9                                    Миллион алых роз  2khK4aqzvb1Dsc2HqaU5zJ   
10                                          For Seoul  6nCKD9aZUCu4gzZ25MzB5n   
11                                    Cranes at Night  5rYPgW2n44YxtJ0qzQAEjK   
12                                             Nevada  02shCNmb6IvgB5jLqKjtkK   
13                                           Netzwerk  5Wxg8ocgiYkLozRjJhJReM   
14                                              Bloom  0xoCyCfPwoRZcaMi7oXzjP   
15                                           Импульсы  3b1EjkgORmChXUcVC63x8w   
16  Goodbye (feat. Soap&Skin) [Theme from Dark, A ...  3wAEhWSx5GvfCkqJ1SPA5S   
17                                Jetlands und Wolken  1Yf98TVe3AOClCCbbQnQjr   
18                                           Mountiri  6x286HQsHph7S4yKcgINow   
19                                           Sandburg  3xJLi9PvRToiqb9zRqpTQ7   
20                                             Voices  7e6VKt7DAcvv0lxNMbPESM   
21                            Do Not Do Me (Like Dis)  4DGrOp8YhMx5WocfOd0pQE   
22                                              Hades  04LUmqZOsmanOAqiqnyJ3S   
23                                         Katzengold  7jI02mCFkOSivg3NjomHcG   
24                                          Your Lips  3ByKecIVNswbnmOMCTXPt6   

                     artist  duration  popularity  danceability  energy  key  \
0           Oliver Koletzki    370333          34         0.836   0.495    9   
1                    Kisnou    211200          41         0.467   0.655    4   
2                     Gidge    484682          49         0.592   0.413    6   
3          Babek Mamedrzaev    177968          60         0.790   0.922    7   
4            Thomas Lizzara    340505          35    

In [52]:
recomm_dfs = []
for i in range(5,26,5):
    recomms = sp_m.recommendations(seed_tracks = seed_tracks[i-5:i],limit = 25)
    recomms_df = append_audio_features(create_df_recommendations(recomms),sp_m)
    recomm_dfs.append(recomms_df)
recomms_df = pd.concat(recomm_dfs)

In [53]:
recomms_df.reset_index(drop = True, inplace= True)

In [54]:
recomms_df

track_name  \
0                                       Middle Finger   
1                                            Ненавижу   
2                                            Drifting   
3                             Do Not Do Me (Like Dis)   
4                                          Hypnotized   
..                                                ...   
120                           Oloro Nyager - Club Mix   
121                 Stickin' (feat. Masego & VanJess)   
122                                       Eyes Closed   
123  Baptize (with JID & EARTHGANG feat. Ant Clemons)   
124                                          Palomita   

                                 album                track_id  \
0               Middle Finger - Single  0IK8y0vAO0sM630HsxwLht   
1                             Ненавижу  2dxCRx56x3hNzuZfnGhzL7   
2                             Drifting  3GWomAWy6PxtyzlfqKhTIl   
3              Do Not Do Me (Like Dis)  4DGrOp8YhMx5WocfOd0pQE   
4                     Großstadtmärchen  0V2GX0aukyZMt6nSMxfOJk   
..                                 ...                     ...   
120                       Oloro Nyager  4cPqGklG4Ffb7AVlWhuWQt   
121  Stickin' (feat. Masego & VanJess)  24KUvSg9QsX6FWsOmN0ZxP   
122                   Moments Of Truth  5AyqWYv3gHsvQ0FJuasFs9   
123                         Spilligion  5zWOqc9si4XnemdxZH4WGG   
124                              Endup  5xWzVhDpu6hpcoiO2euL8u   

                artist  duration  popularity  danceability  energy  key  \
0               Bohnes    229986          70         0.685   0.656    1   
1         Misha Marvin    186000          36         0.837   0.517    1   
2            TWO LANES    187000          46         0.592   0.719    7   
3          Moonbootica    194081          26         0.774   0.818    8   
4      Oliver Koletzki    340560          58         0.764   0.381    7   
..                 ...       ...         ...           ...     ...  ...   
120  Djuma Soundsystem    424918          25         0.797   0.694    4   
121     Sinead Harnett    188805          64         0.668   0.450    8   
122     Tim Engelhardt    216784          23         0.835   0.373    6   
123   Spillage Village    293873          65         0.669   0.667    6   
124                 Nu    483824          48         0.833   0.541    6   

     loudness  mode  speechiness  acousticness  instrumentalness  liveness  \
0      -6.013     0       0.0383       0.43700          0.000000    0.3470   
1      -6.316     1       0.0449       0.00915          0.000008    0.6460   
2      -5.781     1       0.0757       0.02060          0.616000    0.2210   
3      -6.445     1       0.0502       0.01820          0.012600    0.0884   
4     -11.845     1       0.0535       0.04580          0.714000    0.1800   
..        ...   ...          ...           ...               ...       ...   
120   -10.212     1       0.0473       0.03100          0.847000    0.6680   
121    -8.257     0       0.3140       0.07220          0.000000    0.1090   
122   -12.303     0       0.0602       0.22800          0.620000    0.1180   
123    -7.028     0       0.3620       0.18700          0.000000    0.1720   
124   -13.532     1       0.0709       0.12000          0.687000    0.0903   

     valence    tempo  
0      0.392  128.003  
1      0.222  119.986  
2      0.207  159.876  
3      0.434  119.009  
4      0.298  122.988  
..       ...      ...  
120    0.415  122.006  
121    0.633   82.999  
122    0.208  115.004  
123    0.512   77.985  
124    0.160  105.014  

[125 rows x 17 columns]

The 125 recommendations are further filtered by their similarity to the known tracks in the playlist.

In [56]:
similarity_score = create_similarity_score(new_playlist_df,recomms_df)

In [58]:
new_playlist_df.shape

(25, 17)

In [59]:
recomms_df

track_name  \
0                                       Middle Finger   
1                                            Ненавижу   
2                                            Drifting   
3                             Do Not Do Me (Like Dis)   
4                                          Hypnotized   
..                                                ...   
120                           Oloro Nyager - Club Mix   
121                 Stickin' (feat. Masego & VanJess)   
122                                       Eyes Closed   
123  Baptize (with JID & EARTHGANG feat. Ant Clemons)   
124                                          Palomita   

                                 album                track_id  \
0               Middle Finger - Single  0IK8y0vAO0sM630HsxwLht   
1                             Ненавижу  2dxCRx56x3hNzuZfnGhzL7   
2                             Drifting  3GWomAWy6PxtyzlfqKhTIl   
3              Do Not Do Me (Like Dis)  4DGrOp8YhMx5WocfOd0pQE   
4                     Großstadtmärchen  0V2GX0aukyZMt6nSMxfOJk   
..                                 ...                     ...   
120                       Oloro Nyager  4cPqGklG4Ffb7AVlWhuWQt   
121  Stickin' (feat. Masego & VanJess)  24KUvSg9QsX6FWsOmN0ZxP   
122                   Moments Of Truth  5AyqWYv3gHsvQ0FJuasFs9   
123                         Spilligion  5zWOqc9si4XnemdxZH4WGG   
124                              Endup  5xWzVhDpu6hpcoiO2euL8u   

                artist  duration  popularity  danceability  energy  key  \
0               Bohnes    229986          70         0.685   0.656    1   
1         Misha Marvin    186000          36         0.837   0.517    1   
2            TWO LANES    187000          46         0.592   0.719    7   
3          Moonbootica    194081          26         0.774   0.818    8   
4      Oliver Koletzki    340560          58         0.764   0.381    7   
..                 ...       ...         ...           ...     ...  ...   
120  Djuma Soundsystem    424918          25         0.797   0.694    4   
121     Sinead Harnett    188805          64         0.668   0.450    8   
122     Tim Engelhardt    216784          23         0.835   0.373    6   
123   Spillage Village    293873          65         0.669   0.667    6   
124                 Nu    483824          48         0.833   0.541    6   

     loudness  mode  speechiness  acousticness  instrumentalness  liveness  \
0      -6.013     0       0.0383       0.43700          0.000000    0.3470   
1      -6.316     1       0.0449       0.00915          0.000008    0.6460   
2      -5.781     1       0.0757       0.02060          0.616000    0.2210   
3      -6.445     1       0.0502       0.01820          0.012600    0.0884   
4     -11.845     1       0.0535       0.04580          0.714000    0.1800   
..        ...   ...          ...           ...               ...       ...   
120   -10.212     1       0.0473       0.03100          0.847000    0.6680   
121    -8.257     0       0.3140       0.07220          0.000000    0.1090   
122   -12.303     0       0.0602       0.22800          0.620000    0.1180   
123    -7.028     0       0.3620       0.18700          0.000000    0.1720   
124   -13.532     1       0.0709       0.12000          0.687000    0.0903   

     valence    tempo  
0      0.392  128.003  
1      0.222  119.986  
2      0.207  159.876  
3      0.434  119.009  
4      0.298  122.988  
..       ...      ...  
120    0.415  122.006  
121    0.633   82.999  
122    0.208  115.004  
123    0.512   77.985  
124    0.160  105.014  

[125 rows x 17 columns]

In [60]:
[np.argmax(i) for i in similarity_score]

[124,
 23,
 44,
 111,
 114,
 98,
 121,
 32,
 42,
 68,
 12,
 101,
 84,
 117,
 54,
 54,
 76,
 89,
 109,
 96,
 99,
 103,
 93,
 100,
 63]

In [61]:
final_recomms=recomms_df.loc[[np.argmax(i) for i in similarity_score]]
final_recomms = final_recomms.drop_duplicates()

In [62]:
new_playlist_df = new_playlist_df.append(final_recomms)

In [63]:
new_playlist_df = new_playlist_df.drop_duplicates()
new_playlist_df.reset_index(drop = True, inplace = True)

# The playlist is finished!

Now the only thing left to do is to add the tracks to the playlist.

(Adding a nice picture and thanking your girlfriend for her patience in the playlist description are **not optional**)

In [64]:
new_playlist_df

track_name  \
0                              Huldra - Other Version   
1                                          Дни и ночи   
2                                             Phoenix   
3                                              Zunder   
4                                          By My Side   
5                               Wenn der Winter kommt   
6                                         Rolls Royce   
7   Points of Authority / 99 Problems / One Step C...   
8                                      WAIDMANNS HEIL   
9                                            Handsome   
10                                  Echo of the Woods   
11                                    EIN LETZTES MAL   
12                                    ABER WO BIST DU   
13                                          Zim Zimma   
14                Lass Das Licht An (feat. Deichkind)   
15                                      Lass Sie Gehn   
16                      Aufstehn! (feat. CeeLo Green)   
17                     Risky Business - Mathame Remix   
18                                      Berlin Nights   
19                                 Stargazing - Remix   
20                                      The Beginning   
21        When The Interlude Sounds Like Falling Rain   
22                                       Mind Doodles   
23                                            Fragile   
24                   Mystery Lady (feat. Don Toliver)   
25                                           Palomita   
26                                  We Want Your Soul   
27                   Run to the Hills - 2015 Remaster   
28                                           Rim Jhim   
29                                       Need You Now   
30                           Sun In Your Eyes - Mixed   
31                  Stickin' (feat. Masego & VanJess)   
32                                 Twisted Transistor   
33                                    Electric Guitar   
34                                       Will - Remix   
35                                               Dusk   
36                         Heart Don't Stand a Chance   
37                     Lass sie tanzen (Square Dance)   
38                                      Nottamun Town   
39               Wow Freestyle (feat. Kendrick Lamar)   
40                                         To The Top   
41                      Desire - Solomun Remix (Edit)   
42                                      Searching For   
43                                          Versatile   
44                                               Jein   
45                                    Time Of The G's   
46                                     Großstadtwüste   
47                                              Jafar   
48                                         Home Again   

                                      album                track_id  \
0                                    Huldra  569gNjph2g07MmjtMm6vKm   
1                                Дни и ночи  3A88HuZw5yiZOAcon4RkZk   
2                                   Phoenix  6zAiRKvAMlXHxEtyO4yxIO   
3                               Bittersweet  10Sp3ZHJUkSoYCN9NZO7QL   
4                           Ancient Shadows  0eSWvkrCADGx0409mebsP4   
5                      Mittelpunkt der Welt  6aj2zW1HmhZNQoeGR1avif   
6                               Rolls Royce  22XFe65IH0P2RY0uowBqdI   
7         Collision Course (Deluxe Version)  65eohvrL4ttjA7EfFkQOhX   
8   LIEBE IST FÜR ALLE DA (SPECIAL EDITION)  6ey8jr96hlSORKyeJk9d8d   
9                         Heavy Is The Head  766N0mjf1KzAAYsA1eAJaN   
10                        Echo of the Woods  6d8mjJCtnlFtgGv3erEs7c   
11                                    NACHT  04EJyZSlhPPfFOo1NRn2vl   
12                                    NACHT  2dGtLJysTSI9cbQ6TulL8V   
13                                Evolution  17nPeSliosCi427f0lUb75   
14                                  BAM BAM  0BJ5N5yfEPBAimh9LprQvc   
15                                  BAM BAM  4RYKr1R3tXr

In [65]:
new_playlist_df["track_id"].tolist()

['569gNjph2g07MmjtMm6vKm',
 '3A88HuZw5yiZOAcon4RkZk',
 '6zAiRKvAMlXHxEtyO4yxIO',
 '10Sp3ZHJUkSoYCN9NZO7QL',
 '0eSWvkrCADGx0409mebsP4',
 '6aj2zW1HmhZNQoeGR1avif',
 '22XFe65IH0P2RY0uowBqdI',
 '65eohvrL4ttjA7EfFkQOhX',
 '6ey8jr96hlSORKyeJk9d8d',
 '766N0mjf1KzAAYsA1eAJaN',
 '6d8mjJCtnlFtgGv3erEs7c',
 '04EJyZSlhPPfFOo1NRn2vl',
 '2dGtLJysTSI9cbQ6TulL8V',
 '17nPeSliosCi427f0lUb75',
 '0BJ5N5yfEPBAimh9LprQvc',
 '4RYKr1R3tXrITqY1zWiTNi',
 '2KpvM181ZwtIWwIElZx4vI',
 '5R4hprpCcdgKz1DsPoh9p2',
 '6gG1R1bFdJeNc2ERAwXxCb',
 '1HK6TWeHG9q5upt2WcI0sw',
 '4njhWDhTAjhReWtYkiMH9t',
 '5lWVqc9kPplvlDmLtFls02',
 '4ZVqqf3Eo4uXSeFUrYD5lw',
 '59g3Adj1Vdhja52OrhURdE',
 '2Ee9amLUslOwgoJWZEpSSD',
 '5xWzVhDpu6hpcoiO2euL8u',
 '5Wxg8ocgiYkLozRjJhJReM',
 '4Zc7TCHzuNwL0AFBlyLdyr',
 '7F3q8BR1VVJDrDjGRi5byr',
 '07poLV9zNLuakMGUYTXvyZ',
 '1U5vvKgK8pVT0dpwY2CWRh',
 '24KUvSg9QsX6FWsOmN0ZxP',
 '77VW8u6inET54YAKN1RSnd',
 '7ymr9kOckM8Uw6qNs7My1W',
 '6WI33r2QRxyxttn3KH0XkC',
 '0BKHgXSyG6kquLakA4yNVB',
 '25GlFJq5QNAXyVgJvCZ4Mf',
 

**Note:** Here I am using `user_playlist_add_tracks()` to **add** to an existing playlist. It is possible to create one from scratch, however it wasn't necessary here. 

In the script I am using `playlist_replace_items()` as I don't just want new songs to be added but old ones to be deleted.

In [66]:
sp_m.user_playlist_add_tracks(usernames[0],
                              playlist_id="spotify:playlist:1Vcqtv3nE7QOJ4KFvK7bT8",
                              tracks = new_playlist_df["track_id"].tolist())

{'snapshot_id': 'MTMsODlmMGU0MGFjNGU1MGQ4OWUzZTc2M2ZiYmRlZTgxMWNkZmNmOTA2Mg=='}

In [67]:
new_playlist_df.to_csv(path/"Playlist.csv")